### PR TITLE
PrepareMergeSupportForChangeFeedProcessor

### PR DESCRIPTION
**Pull request by xinlian12**
# Description
Prepare changeFeedProcessor for merge support. Main item tracked by: https://github.com/Azure/azure-sdk-for-java/issues/32131

From high level overview, separate internal changeFeedProcessor into pkversion and epkversion.
`pkversion` -> use partitionKeyRangeBased lease
`epkversion` -> use epkRangeBased lease

**Note:** Only `epkversion` will support merge due to the lease requirements.
**public API mapping:**
`handleChanges` -> `pkversion`
`handleAllVersionsAndDeletesChanges` -> `epkversion`
`handleLatestVersionChanges` -> `epkversion`

# Major Changes included
- All files under `incremental` folder is moved to `pkversion`
- All files under `fullFidelity` folder is moved to `epkversion`
- `FeedRangeGoneHandler`, `FeedRangeGoneMergeHandler`, `FeedRangeGoneSplitHandler` added for merge/split handling
- Others to documented

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/v1/feedRangeGoneHandler/FeedRangeGoneSplitHandler.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+                .doOnComplete(() -> {
+                    logger.info(
+                            "Lease with token {} split into {}",
+                            this.lease.getLeaseToken(),
+                            StringUtils.join(leaseTokens, ","));
+                });
+    }
+
+
+    private String getEffectiveChildLeaseContinuationToken(FeedRangeEpkImpl childLeaseFeedRange, String parentLeaseCT) {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/v1/feedRangeGoneHandler/FeedRangeGoneSplitHandler.java:95]: @FabianMeiswinkel/ @kushagraThapar can you help double check the logic here? when split happens, I think we should reconstruct the continuationToken with the token(the lsn) from parent lease, is this the right way?

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/v1/feedRangeGoneHandler/FeedRangeGoneSplitHandler.java:95]: Yes - that is the correct way for splits.

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/v1/feedRangeGoneHandler/FeedRangeGoneSplitHandler.java:95]: If you have any follow-up questions, there Matias has all the context - he implemented the merge support in .Net CFP and merge support in general for .Net - so can definitely help review anything if you need help there.

**Review by FabianMeiswinkel — APPROVED** [review]
Thanks - really good refactoring. Besides helping to prepare for merge support it makes the code a lot more readable and more intutive to understand.

File: sdk/cosmos/azure-cosmos/src/samples/java/com/azure/cosmos/ChangeFeedProcessorLatestVersionChangesCodeSnippet.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
+package com.azure.cosmos;
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/samples/java/com/azure/cosmos/ChangeFeedProcessorLatestVersionChangesCodeSnippet.java:1]: missing license header.

**Review by kushagraThapar — APPROVED** [review]
Great work @xinlian12, Added few suggestions. 
Wondering if we should rename v0 -> `pkversion` and v1 -> `epkversion`

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/common/ExceptionClassifier.java [hunk 3]
```diff
@@ -27,8 +27,9 @@ public static StatusCodeErrorType classifyClientException(CosmosException client
             return StatusCodeErrorType.PARTITION_NOT_FOUND;
         }
 
-        if (clientException.getStatusCode() == ChangeFeedHelper.HTTP_STATUS_CODE_GONE && (subStatusCode == SubStatusCode_PartitionKeyRangeGone || subStatusCode == SubStatusCode_Splitting)) {
-            return StatusCodeErrorType.PARTITION_SPLIT;
+        if (clientException.getStatusCode() == ChangeFeedHelper.HTTP_STATUS_CODE_GONE
+                && (subStatusCode == SubStatusCode_PartitionKeyRangeGone || subStatusCode == SubStatusCode_Splitting_Or_Merging)) {
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/common/ExceptionClassifier.java:31]: Yes, they do share the same statusCode and substatusCode

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/ChangeFeedProcessorBuilder.java [hunk 4]
```diff
@@ -117,8 +124,39 @@ public ChangeFeedProcessorBuilder leaseContainer(CosmosAsyncContainer leaseConta
...
+     *     &#125;
+     * &#125;&#41;
+     * </pre>
+     * <!-- end com.azure.cosmos.latestVersionChanges.handleChanges -->
+     *
+     * @param consumer the {@link Consumer} to call for handling the feeds.
+     * @return current Builder.
+     */
+    @Beta(value = Beta.SinceVersion.V4_38_0, warningText = Beta.PREVIEW_SUBJECT_TO_CHANGE_WARNING)
+    public ChangeFeedProcessorBuilder handleLatestVersionChanges(Consumer<List<JsonNode>> consumer) {
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/ChangeFeedProcessorBuilder.java]: > Great work @xinlian12, Added few suggestions. Wondering if we should rename v0 -> `pkversion` and v1 -> `epkversion`

Thanks @kushagraThapar. Have updated to use `epkRangeVersion` and `pkRangeVersion`  instead

[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/ChangeFeedProcessorBuilder.java]: > Do we plan to expose `JsonNode` as part of this? How about `ChangeFeedProcessorItem` ?

As discussed offline, I have removed the public API change in this PR - will add them back when the migration piece code is implemented

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/common/ChangeFeedContextClientImpl.java [hunk 5]
```diff
@@ -82,6 +81,33 @@ public void setScheduler(Scheduler scheduler) {
...
+                    return clientWrapper.getPartitionKeyRangeCache().tryGetOverlappingRangesAsync(
+                            null,
+                            collection.getResourceId(),
+                            range,
+                            true,
+                            null);
+                })
+                .flatMap(pkRangesValueHolder -> {
+                    if (pkRangesValueHolder == null || pkRangesValueHolder.v == null) {
+                        logger.warn("There is no overlapping ranges found for range {}", range);
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/common/ChangeFeedContextClientImpl.java]: updated

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/exceptions/FeedRangeGoneException.java [hunk 6]
```diff
@@ -5,25 +5,25 @@
 /**
  * Exception occurred during partition split.
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/exceptions/FeedRangeGoneException.java]: updated

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/v0/PartitionSynchronizer.java [hunk 7]
```diff
@@ -1,7 +1,6 @@
-// Copyright (c) Microsoft Corporation. All rights reserved.
-// Licensed under the MIT License.
-package com.azure.cosmos.implementation.changefeed;
+package com.azure.cosmos.implementation.changefeed.v0;
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/v0/PartitionSynchronizer.java]: added

File: sdk/cosmos/azure-cosmos/src/samples/java/com/azure/cosmos/ChangeFeedProcessorLatestVersionChangesCodeSnippet.java (hunk 2 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/samples/java/com/azure/cosmos/ChangeFeedProcessorLatestVersionChangesCodeSnippet.java]: As discussed offline, I have removed all the public API changes, so this file is also removed

**Comment by xinlian12** [comment]
Run the following tests locally and passed:

readAllItems
createItem_withBulk
before_ParallelDocumentQueryTest
before_ReadFeedStoredProceduresTest
before_UserDefinedFunctionQueryTest
createRecoversFrom410GoneClientGeneratedOnPartitionSplitDuringIdleTime

**Pull request merged**